phoenix-core-client/src/main/java/org/apache/phoenix/util/json/BsonDataFormat.java

```diff
@@ -0,0 +1,201 @@
+import com.jayway.jsonpath.PathNotFoundException;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.bson.BsonBinaryReader;
+import org.bson.BsonDocument;
+import org.bson.BsonDocumentReader;
+import org.bson.BsonValue;
+import org.bson.RawBsonDocument;
+import org.bson.codecs.BsonDocumentCodec;
+import org.bson.codecs.DecoderContext;
+import org.bson.codecs.RawBsonDocumentCodec;
+import org.bson.io.ByteBufferBsonInput;
+
+import java.nio.ByteBuffer;
+import java.sql.Types;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class BsonDataFormat implements JsonDataFormat {
+    @Override
+    public byte[] toBytes(Object object) {
+        return Bytes.toBytes(((RawBsonDocument) object).getByteBuffer().asNIO());
+    }
+
+    @Override
+    public Object toObject(String value) {
+        return RawBsonDocument.parse(value);
+    }
+
+    @Override
+    public Object toObject(byte[] bytes, int offset, int length) {
+        return new RawBsonDocument(bytes, offset, length);
+    }
+
+    @Override
+    public int estimateByteSize(Object o) {
+        RawBsonDocument rawBSON = (RawBsonDocument) o;
+        return rawBSON.size();
+    }
+
+    @Override
+    public int getValueType(Object obj, String jsonPathExprStr) {
+        BsonValue value = getBsonValue(jsonPathExprStr, (RawBsonDocument) obj);
+        return getSqlType(value);
+    }
+
+    @Override
+    public Object getValue(Object obj, String jsonPathExprStr) {
+        BsonValue value = getBsonValue(jsonPathExprStr, (RawBsonDocument) obj);
+        return getValue(value);
+    }
+
+    private Object getValue(BsonValue value) {
+        if (value != null) {
+            switch (value.getBsonType()) {
+            case INT32:
+                return value.asInt32().getValue();
+            case INT64:
+                return value.asInt64().getValue();
+            case STRING:
+            case SYMBOL:
+                return value.asString().getValue();
+            case DECIMAL128:
+                return value.asDecimal128().doubleValue();
+            case DOUBLE:
+                return value.asDouble().getValue();
+            case BOOLEAN:
+                return value.asBoolean().getValue();
+            case BINARY:
+                return value.asBinary().getData();
+            case DATE_TIME:
+                return value.asDateTime().getValue();
+            case DOCUMENT:
+                return value.asDocument().toJson();
+            case ARRAY:
+                return readArray(value).toString();
+            default:
+                return null;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public ByteBuffer updateValue(Object top, String jsonPathExprStr, String newVal) {
+        Configuration conf = Configuration.builder().jsonProvider(new BsonJsonProvider()).build();
+        BsonValue newValue = JsonPath.using(conf).parse(newVal).json();
+        BsonDocument root = fromRaw((RawBsonDocument) top);
```

Comment:
Do we need to use fromRaw() here which uses BsonBinaryReader and then decode RawBsonDocument to BsonDocument?
Can we use this directly? It seems safe type cast.

```
BsonDocument root = ((RawBsonDocument) top).asDocument()
```

Reply:
The reason we use RawBSONDocument is because we want to do lazy parsing of bson bytes. Once we have the byte buffer representing json if we call `asDocument` it will parse the entire byte buffer and construct a BsonDocument. But then you have lost the efficiency gains. We only want to parse the byte[] which is queried in the jsonpath expression leaving the rest of the bytes uninterpreted. By uninterpreted I mean we don't parse those bytes  but just read them into memory into the buffer. This is where the performance gains come from. If you parse everything you lose that.

Reply:
Sounds good.

@ranganathg could you document this as comments for `fromRaw()`? 
`fromRaw()` has only one comment `// Transform to an in memory BsonDocument instance` as of now.

Reply:
Btw the definition of `asDocument()`

```
  /**
     * Gets this value as a BsonDocument if it is one, otherwise throws exception
     *
     * @return a BsonDocument
     * @throws org.bson.BsonInvalidOperationException if this value is not of the expected type
     */
    public BsonDocument asDocument() {
        throwIfInvalidType(BsonType.DOCUMENT);
        return (BsonDocument) this;
    }
```

Reply:
The problem here is that, `asDocument()` just does the typecast, so even if we use it, the underlying object is still of Raw type, meaning we will not be able to update the document. Hence, we need to create updatable `BsonDocument` object using `BsonBinaryReader`.

Reply:
Will take this another iteration